crosscat/utils/inference_utils.py

```diff
@@ -124,30 +118,45 @@ def calculate_MI_bounded_discrete(X, Y, M_c, X_L, X_D):
         component_models_X[i] = cluster_models[X]
         component_models_Y[i] = cluster_models[Y]
 
-    MI = 0.0
+    def marginal_predictive_by_cluster(value, component_models):
+        return numpy.array([
+            component_models[j].calc_element_predictive_logp(value)
+            + cluster_logps[j]
+            for j in range(n_clusters)])
+
+    x_marginal_predictives_by_cluster = \
+        [marginal_predictive_by_cluster(x, component_models_X)
+         for x in x_values]
+
+    # \sum_c P(x|c)P(c)
+    x_net_marginal_predictives = \
+        [logsumexp(ps) for ps in x_marginal_predictives_by_cluster]
+
+    y_marginal_predictives_by_cluster = \
+        [marginal_predictive_by_cluster(y, component_models_Y)
+         for y in y_values]
 
-    for x in x_values:
-        for y in y_values:
-             # calculate marginal logs
-            Pxy = numpy.zeros(n_clusters)   # P(x,y), Joint distribution
-            Px = numpy.zeros(n_clusters)    # P(x)
-            Py = numpy.zeros(n_clusters)    # P(y)
+    # \sum_c P(y|c)P(c)
+    y_net_marginal_predictives = \
+        [logsumexp(ps) for ps in y_marginal_predictives_by_cluster]
 
-            # get logp of x and y in each cluster. add cluster logp's
-            for j in range(n_clusters):
+    MI = 0.0
 
-                Px[j] = component_models_X[j].calc_element_predictive_logp(x)
-                Py[j] = component_models_Y[j].calc_element_predictive_logp(y)
-                Pxy[j] = Px[j] + Py[j] + cluster_logps[j]   # \sum_c P(x|c)P(y|c)P(c), Joint distribution
-                Px[j] += cluster_logps[j]                   # \sum_c P(x|c)P(c)
-                Py[j] += cluster_logps[j]                   # \sum_c P(y|c)P(c)
+    for (i,x) in enumerate(x_values):
+        x_marginals = x_marginal_predictives_by_cluster[i]
+        for (j,y) in enumerate(y_values):
+            y_marginals = y_marginal_predictives_by_cluster[j]
+            # cluster prob is double-counted in sum of marginals
+            joint_predictive_by_cluster = \
+                x_marginals + y_marginals - cluster_logps
 
-            # sum over clusters
-            Px = logsumexp(Px)
-            Py = logsumexp(Py)
-            Pxy = logsumexp(Pxy)
+            # \sum_c P(x|c)P(y|c)P(c), Joint distribution
+            joint_predictive = logsumexp(joint_predictive_by_cluster)
 
-            MI += numpy.exp(Pxy)*(Pxy - (Px + Py))
+            MI += math.exp(joint_predictive) * \
```

Comment:
Interestingly, using math.exp rather than numpy.exp there seems to make that whole function about 10% faster on my test workload (64 models, 500 clusters between them, 54 values in each categorical).  Presumably this is because numpy.exp has all sorts of machinery for looping over numpy arrays in it.

Suggestions?  Shall we define ieee_exp in a utility file and replace all uses of math.exp with it?  (I would rather do that as a separate change list).